libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/helpers/DataClass.kt

```diff
@@ -62,17 +65,193 @@ fun TypeSpec.Builder.makeDataClassFromProperties(properties: List<PropertySpec>)
+        .addStatement("return true")
+        .endControlFlow()
+        .beginControlFlow("if (other is %T)", className)
+        .apply {
+          if (propertySpecs.isEmpty()) {
+            add("return true")
+          } else {
+            add("return %L", propertySpecs
+                .excludeInternalProperties()
+                .map { equalsCode(it) }
+                .joinToCode("&& ")
+            )
+          }
+        }
+        .endControlFlow()
+        .addStatement("return false")
+        .build()
+  }
+
+  return addFunction(FunSpec.builder(Identifier.equals)
+      .addModifiers(KModifier.OVERRIDE)
+      .addParameter("other", KotlinSymbols.Any.copy(nullable = true))
+      .returns(KotlinSymbols.Boolean)
+      .addCode(methodCode(build().name))
+      .build())
+}
+
+fun TypeSpec.Builder.withHashCodeImplementation(): TypeSpec.Builder = apply {
+  fun hashPropertyCode(property: PropertySpec) =
+      CodeBlock.builder()
+          .addStatement("${Identifier.__h} *= 31")
+          .addStatement("${Identifier.__h} += %L.hashCode()", property.name)
+          .build()
+
+  fun methodCode(): CodeBlock {
+    if (propertySpecs.isEmpty()) {
+      return CodeBlock.builder().addStatement("return·this::class.hashCode()").build()
+    }
+    return CodeBlock.builder()
+        .beginControlFlow("if (%L == null)", MEMOIZED_HASH_CODE_VAR)
+        .addStatement("var ${Identifier.__h} = %L.hashCode()", propertySpecs.getOrNull(0)?.name ?: "null")
+        .add(propertySpecs
+            .drop(1)
+            .excludeInternalProperties()
+            .map(::hashPropertyCode)
+            .fold(CodeBlock.builder(), CodeBlock.Builder::add)
+            .build())
+        .addStatement("%L = ${Identifier.__h}", MEMOIZED_HASH_CODE_VAR)
+        .endControlFlow()
+        .addStatement("return %L!!", MEMOIZED_HASH_CODE_VAR)
+        .build()
+  }
+
+  check(propertySpecs.all { !it.mutable }) {
+    "Cannot generate hashCode function for class with mutable properties, memoization would fail"
+  }
+
+  if (propertySpecs.isNotEmpty()) {
+    addProperty(
+        PropertySpec.builder(MEMOIZED_HASH_CODE_VAR, KotlinSymbols.Int.copy(nullable = true), KModifier.PRIVATE)
+            .mutable()
+            .initializer("null")
+            .build()
+    )
+  }
+
+  addFunction(FunSpec.builder(Identifier.hashCode)
+      .addModifiers(KModifier.OVERRIDE)
+      .returns(KotlinSymbols.Int)
+      .addCode(methodCode())
+      .build()
+  )
+}
+
+fun TypeSpec.Builder.withToStringImplementation(): TypeSpec.Builder {
+  val name = build().name ?: return this
+  fun printPropertiesTemplate() =
+      "$name(" + propertySpecs
+            .excludeInternalProperties()
+            .joinToString(",") { "${it.name}=\$${it.name}" } + ")"
+
+  fun methodCode() =
+      CodeBlock.builder()
+          .beginControlFlow("if (%L == null)", MEMOIZED_TO_STRING_VAR)
+          .add("%L = %P", MEMOIZED_TO_STRING_VAR, printPropertiesTemplate())
```

Comment:
I'm not sure how the performance of this compares to Kotlin's implementation which uses `irConcat`.  https://github.com/JetBrains/kotlin/blob/master/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/DataClassMembersGenerator.kt#L207